dynamic/internal/shim/loader.go

```diff
@@ -0,0 +1,263 @@
+	for k, packages := range systemCache.AllAvailablePackages() {
+		if k != addr {
+			continue
+		}
+		// packages is sorted by precedence, so the first cached result is safe to
+		// use.
+		acceptable := versions.MeetingConstraints(version)
+		for _, p := range packages {
+			if acceptable.Has(p.Version) {
+				slog.InfoContext(ctx, "Found cached provider",
+					slog.Any("addr", addr.String()),
+					slog.Any("version", p.Version.String()))
+				p := p
+				return runProvider(ctx, &p)
+			}
+		}
+	}
+
+	// We have not found a package that fits our constraints, so we need to download
+	// one.
+
+	source := getproviders.NewRegistrySource(disco.New())
+
+	availableVersions, warnings, err := source.AvailableVersions(ctx, addr)
+	if err != nil {
+		// TODO Handle error kinds with distinct error messages
+		return nil, err
+	}
+
+	// TODO: Don't ignore warnings
+	ignore(warnings)
+
+	desiredVersion := availableVersions.NewestInSet(versions.MeetingConstraints(version))
+	if desiredVersion == versions.Unspecified {
+		return nil, fmt.Errorf("Could not resolve a version from %s: %s", addr, version)
+	}
+
+	meta, err := source.PackageMeta(ctx, addr, desiredVersion, getproviders.CurrentPlatform)
+	if err != nil {
+		return nil, err
+	}
+
+	_, err = systemCache.InstallPackage(ctx, meta, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	p := systemCache.ProviderVersion(addr, desiredVersion)
+	contract.Assertf(p != nil, "We just downloaded (%s,%s) so it should be in the cache", addr, desiredVersion)
+
+	return runProvider(ctx, p)
+}
+
+// runProvider produces a provider factory that runs up the executable
+// file in the given cache package and uses go-plugin to implement
+// providers.Interface against it.
+func runProvider(ctx context.Context, meta *providercache.CachedProvider) (Provider, error) {
+	execFile, err := meta.ExecutableFile()
+	if err != nil {
+		return nil, err
+	}
+
+	config := &plugin.ClientConfig{
+		HandshakeConfig:  tfplugin.Handshake,
+		Logger:           logging.NewProviderLogger(""),
+		AllowedProtocols: []plugin.Protocol{plugin.ProtocolGRPC},
+		Managed:          true,
+		Cmd:              exec.Command(execFile),
+		AutoMTLS:         true,
+		VersionedPlugins: tfplugin.VersionedPlugins,
+		SyncStdout:       logging.PluginOutputMonitor(fmt.Sprintf("%s:stdout", meta.Provider)),
+		SyncStderr:       logging.PluginOutputMonitor(fmt.Sprintf("%s:stderr", meta.Provider)),
+	}
+
+	client := plugin.NewClient(config)
+	rpcClient, err := client.Client()
+	if err != nil {
+		return nil, err
+	}
+
+	raw, err := rpcClient.Dispense(tfplugin.ProviderPluginName)
+	if err != nil {
+		return nil, err
+	}
+
+	// store the client so that the plugin can kill the child process
+	protoVer := client.NegotiatedVersion()
+	switch protoVer {
+	case 5:
+		p := raw.(*tfplugin.GRPCProvider)
+		p.PluginClient = client
+		p.Addr = meta.Provider
+
+		v6, err := tf5to6server.UpgradeServer(ctx, func() tfprotov5.ProviderServer {
```

Comment:
slog a message here we're using a shim to approximate v5 to v6 protocol perhaps.